src/Dictionaries/HierarchyDictionariesUtils.h

```diff
@@ -0,0 +1,468 @@
+
+                    if (unlikely(is_loop))
+                        break;
+
+                    size_t array_element_offset = index.array_element_offset;
+
+                    size_t previous_offset_size = offset > 0 ? offsets[offset - 1] : 0;
+                    size_t start_index = previous_offset_size + array_element_offset;
+                    size_t end_index = offsets[offset];
+
+                    current_hierarchy_depth += end_index - start_index;
+                    elements.insertFromItself(elements.begin() + start_index, elements.begin() + end_index);
+                    break;
+                }
+
+                if (hierarchy_key == hierarchy_null_value || current_hierarchy_depth >= DBMS_HIERARCHICAL_DICTIONARY_MAX_DEPTH)
+                    break;
+
+                already_processes_keys_to_offset[hierarchy_key] = {offsets.size(), current_hierarchy_depth};
+                elements.emplace_back(hierarchy_key);
+                ++current_hierarchy_depth;
+
+                std::optional<KeyType> parent_key = std::forward<GetParentKeyFunc>(get_parent_key_func)(hierarchy_key);
+
+                if (!parent_key.has_value())
+                    break;
+
+                hierarchy_key = *parent_key;
+            }
+
+            offsets.emplace_back(elements.size());
+        }
+
+        ElementsAndOffsets<KeyType> result = {std::move(elements), std::move(offsets)};
+
+        return result;
+    }
+
+    /** Returns array with UInt8 represent if key from in_keys array is in hierarchy of key from keys column.
+      * If value in result array is 1 that means key from in_keys array is in hierarchy of key from
+      * keys array with same index, 0 therwise.
+      * For getting hierarchy implementation uses getKeysHierarchy function.
+      *
+      * Not: keys size must be equal to in_keys_size.
+      */
+    template <typename KeyType, typename IsKeyValidFunc, typename GetParentKeyFunc>
+    PaddedPODArray<UInt8> getIsInHierarchy(
+        const PaddedPODArray<KeyType> & keys,
+        const PaddedPODArray<KeyType> & in_keys,
+        const KeyType & hierarchy_null_value,
+        IsKeyValidFunc && is_key_valid_func,
+        GetParentKeyFunc && get_parent_func)
+    {
+        assert(keys.size() == in_keys.size());
+
+        PaddedPODArray<UInt8> result;
+        result.resize_fill(keys.size());
+
+        detail::ElementsAndOffsets<KeyType> hierarchy = detail::getHierarchy(
+            keys,
+            hierarchy_null_value,
+            std::forward<IsKeyValidFunc>(is_key_valid_func),
+            std::forward<GetParentKeyFunc>(get_parent_func));
+
+        auto & offsets = hierarchy.offsets;
+        auto & elements = hierarchy.elements;
+
+        for (size_t i = 0; i < offsets.size(); ++i)
+        {
+            size_t i_elements_start = i > 0 ? offsets[i - 1] : 0;
+            size_t i_elements_end = offsets[i];
+
+            auto & key_to_find = in_keys[i];
```

Comment:
I am not sure that it works. You have a bunch of keys with getParent function, which form a hierarchy. So, the element from in_keys is in hierarchy if is a parent for some key from `keys`, isn't it?

Reply:
Yes, thats right.